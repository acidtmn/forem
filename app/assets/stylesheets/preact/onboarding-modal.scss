@import 'variables';
@import 'mixins';

#onboarding-container {
  min-height: calc(100vh - 350px);
}

.pages-onboarding {
  min-height: initial;
}

// Modal Box Card
.onboarding-main {
  @include themeable(background-color, theme-container-background, $white);
  @include themeable(border-color, theme-container-border, black);
  align-items: center;
  border-radius: 3px;
  border: 2px solid black;
  box-shadow: 4px 4px black;
  display: flex;
  flex-direction: column;
  font-size: 16px;
  justify-content: center;
  min-height: 460px;
  width: 740px;
  z-index: 1;
}

// Modal Inner Content Spacing
.onboarding-content {
  box-sizing: border-box;
  flex: 1 0 auto;
  padding: 32px 48px 48px;
  width: 100%;

  @media screen and (min-width: 800px) {
  }
}

// Modal Header
.onboarding-content-header {
  margin-bottom: 24px;

  .title,
  .subtitle {
    margin: 0;
  }

  .title {
    font-weight: 900;
    text-align: left;
    margin-bottom: 4px;
  }

  .subtitle {
    color: #363d44;
    font-size: 20px;
    font-weight: 500;
    letter-spacing: normal;
    line-height: 150%;
  }
}

// Modal Navigation
.onboarding-navigation {
  align-self: flex-end;
  flex-shrink: 0;
  width: 100%;
  border-bottom: 2px solid black;

  .navigation-content {
    display: flex;
    justify-content: space-between;
    padding: 16px;
  }
}

// Modal Selection Status
.onboarding-selection-status {
  align-self: flex-end;
  flex-shrink: 0;
  width: 100%;
  border-top: 2px solid black;

  .selection-status-content {
    display: block;
    padding: 16px;
    text-align: center;
  }
}

// Button Styles
.onboarding-body button {
  -webkit-appearance: none;
  background-color: #3b49df;
  border-radius: 3px;
  border: none;
  box-sizing: border-box;
  color: white;
  font-size: 16px;
  font-weight: 500;
  line-height: 150%;
  padding: 0.5em 1.5em;
  text-align: center;
}

// Tag Styles
.tag-container {
  box-sizing: border-box;
  column-gap: 1em;
  display: grid;
  grid-gap: 1em;
  grid-template-columns: repeat(3, 1fr);
  grid-template-rows: auto;
  width: 100%;

  .tag-item {
    background-color: #eef0f1; // smoke-10
    border: 2px solid #eef0f1;
    border-radius: 3px;
    display: grid;
    row-gap: 1em;
    height: 120px;
    padding: 16px;

    &:hover {
      border-color: black;
      background: white;
      cursor: pointer;

      .tag-item-selector {
        display: block;
      }
    }

    .tag-topic-name {
      font-family: 'SFMono-Regular', Consolas, 'Liberation Mono', Menlo, Courier,
        monospace;
      font-size: 16px;
      font-weight: 400;
      margin: 0;
    }

    .tag-item-selector {
      display: none;
      padding: 0;
    }
  }
}

.onboarding-tags {
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  grid-gap: 24px;
  width: 100%;

  &__button {
    font-size: 16px;
    font-weight: 500;
    line-height: 24px;
    padding: 0.5em;
    border: 0;
    margin: 0;
    display: block;
    text-align: center;
    position: absolute;
    left: 16px;
    right: 16px;
    width: calc(100% - 32px);
    bottom: 16px;
    border-radius: 3px;
    transition: all ease-in-out 0.1s;
    transform: translateY(60px);
    overflow: hidden;
    $button: &;

    &-default {
      display: inline-block;
    }

    &-alt {
      display: none;
    }
  }

  &__item {
    border-radius: 3px;
    position: relative;
    overflow: hidden;
    height: 150px;
    $item: &;

    &__inner {
      position: absolute;
      left: 0;
      right: 0;
      top: 0;
      bottom: 0;
      z-index: 2;
      background: #eef0f1;
      transition: all ease-in-out 0.1s;
      padding: 16px;
    }

    &:hover {
      #{$item}__inner {
        background: transparent;
      }

      .onboarding-tags__button {
        transform: translateY(0);
        &:hover {
          transform: translateY(-2px);
        }
      }
    }

    &--selected {
      #{$item}__inner {
        background: transparent;
      }

      .onboarding-tags__button {
        transform: translateY(0);
        $button: &;

        &:hover {
          .onboarding-tags__button-default {
            display: none;
          }

          .onboarding-tags__button-alt {
            display: inline-block;
          }
        }
      }
    }
  }
}

// Onboarding Intro

.onboarding-main.introduction {
  width: 578px;

  .onboarding-content {
    background: #08090a;
    color: white;
    display: flex;
    flex-direction: column;
    justify-content: center;
    padding: 48px;
  }

  figure {
    margin: 0 0 8px 0;

    .sticker-logo {
      height: 48px;
      width: 48px;
      transform: rotate(-30deg);
      padding: 10px;
      vertical-align: middle;
    }
  }

  .introduction-title,
  .introduction-subtitle {
    margin: 0;
  }

  .introduction-title {
    font-size: 48px;
    font-weight: 900;
    line-height: 125%;
  }

  .introduction-subtitle {
    font-size: 24px;
    font-weight: 400;
    line-height: 150%;
  }

  .navigation-content {
    padding: 32px;
    justify-content: center;
  }
}

// Next steps

// Background
.onboarding-body {
  background-image: url('/assets/onboarding-background-white.png');
  position: fixed;
  margin-top: 50px;
  bottom: 0;
  height: 100vh;
  width: 100%;
  overflow: hidden;
  display: flex;
  justify-content: center;
  align-items: center;
  line-height: 200%;
  z-index: 11;
  margin: 0;
  -webkit-background-size: cover;
  -moz-background-size: cover;
  -o-background-size: cover;
  background-size: cover;
  /* Center the image */
  background-position: center;
  /* Do not repeat the image */
  background-repeat: no-repeat;

  &:before {
    display: block;
    content: '';
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    background: rgba(0, 0, 0, 0.65);
    height: 100%;
    width: 100%;
  }

  a {
    @include themeable(color, theme-anchor-color, #2e3b9e);
  }

  .background-image {
    z-index: -1;
    position: absolute;
    top: 60px;
    left: 0px;
  }

  .badge-image {
    position: absolute;
  }

  .onboarding-what-next {
    display: flex;
    flex-direction: column;
    height: 250px;
    overflow: auto;

    @media screen and (min-width: 600px) {
      flex-direction: row;
    }

    a {
      background: #d9e7ff;
      padding: 15px 4px 0px 12px;
      height: 140px;
      margin: 4px auto;
      border-radius: 3px;
      width: 350px;
      display: block;
      line-height: 1.2em;
      font-weight: bold;
      position: relative;
      text-align: center;

      @media screen and (min-width: 600px) {
        height: 220px;
        margin: 4px 4px;
        text-align: left;
        padding: 20px 4px 0px 12px;
      }

      .whatnext-emoji {
        display: none;
        text-align: center;
        position: absolute;
        bottom: 22px;
        left: 0;
        right: 0;
        font-size: 50px;

        @media screen and (min-width: 600px) {
          display: block;
        }
      }
    }
  }

  .onboarding-previous-location {
    text-align: center;
    margin-top: 10px;
    font-weight: bold;
    display: none;

    @media screen and (min-width: 600px) {
      display: block;
    }

    code {
      font-size: 0.6em;
      margin-top: -10px;
      display: block;
      font-weight: 400;
    }
  }

  @media screen and (max-width: 800px) {
    .onboarding-main {
      width: 100%;
      height: 100%;
      background-image: url('/assets/mobile-onboarding-background.png');
      background-size: cover;
      /* Center the image */
      background-position: center;
      /* Do not repeat the image */
      background-repeat: no-repeat;
      border: none;
      color: $black;
    }

    .onboarding-content {
      flex: 0 0 auto;

      h1,
      h2 {
        margin-top: initial;
      }
    }

    span {
      display: block;
    }
  }

  @media screen and (max-width: 650px) {
    .onboarding-main {
      font-size: 18px;
    }

    button {
      width: 100px;
      font-size: 20px;
    }
  }

  @media screen and (max-width: 450px) {
    .onboarding-main {
      font-size: 16px;
    }

    button {
      width: 100px;
      font-size: 18px;
    }
  }

  @media screen and (max-width: 330px) {
    button {
      width: 80px;
      font-size: 14px;
    }

    .sticker-logo {
      height: 40px;
      width: 40px;
    }

    h1 {
      font-size: 20px;
    }
  }
}

// Follow user
.onboarding-body {
  .content-row {
    align-items: center;
    color: #202428; // smoke-90
    display: flex;
    padding: 8px 16px;
    text-align: left;
    width: 100%;

    &:hover {
      background-color: #eef0f1 !important; //smoke-10
    }
  }

  .user-avatar-container {
    margin: 0 16px 0 0;
  }

  .user-avatar {
    border-radius: 100%;
    height: 48px;
    width: 48px;
    object-fit: cover;
    border: 1px solid #eef0f1; //smoke-10
  }

  .user-name,
  .user-summary {
    margin: 0;
  }

  .user-following-status {
    margin-left: 32px;
  }
}
