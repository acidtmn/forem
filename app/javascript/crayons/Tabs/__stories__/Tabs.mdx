# Tabs

## What exactly?
Tabs navigation is supposed to allow users to navigate easily between views within the same context.
Tabs component is responsible only for dipslaying Tabs-like navigation (controls). 
It is **not** supposed to handle displaying tabbed content (literally switching between tab panels).

## Buttons vs Links
Depending on your use case, you may want to either use Buttons or Links for individual tabs. 
Therefore, individual `<Tab />` elements will inherit props for either Buttons or Links, respectively.

## Usage

Nielsen Norman Groups has a great [resource about using tabs correctly](https://www.nngroup.com/articles/tabs-used-right/).
Below you can find couple rules of thumb.

### Short labels
Write short tab labels and use plain language, rather than made-up terms. Tab labels should usually be 1–2 words.
Short labels are more scannable; if you need longer labels, it’s a sign that the choices are too complicated for a tab control.

### Number of tabs

Keep number of tabs limitied. Even though we don't have any defined constraints here, it's probably wise to never exceed 5 or 6 tabs. 
This maintains an uncluttered UI and reduces cognitive load for users. If you think more tabs is needed, then you should yell at your designer.

### Tabs are not inputs

Tabs component should never be considered as Form control, for example acting like radio buttons. 
So don't treat it as another input or part of your Form.

### Provide title

Always provide a desciptive `title` prop for your tabs navigation as this helps screen reader users get to the right one quickly and easily.
Keep in mind, it will not be **visible** in the interface.

### Provide current state

Individual `<Tab />` can have `current` prop which defines currently selected (active) tab. Always have **one** `<Tab current />` at a time.
This will not only apply some extra styling to differentiate currently selected tab but also use appropriate attributes to improve accessibility.

Rememeber that "**current**" (aka *active*, aka *selected*) is defined differently:

- For JSX add `current` prop to appropriate tab.
- For HTML:
    - For `<button>` tabs you need to use `aria-pressed="true"` attribute for *current* tab and `aria-pressed="false"` for the rest.
    - For `<a>` tabs you only need to use `aria-current="page"` attribute for *current* tab.

You should be able to just use code generated in Storybook to get an idea of appropriate syntax for each case.

### JSX

In JSX world you only need to define your `elements` prop to be `buttons` or `links` and appropriate code will be handled automatically.

- For `buttons`, individual `<Tab />` elements will turn into `<Button />`. Keep in mind to add `onClick` attribute (or anything else that will handle actual navigation).
- For `links`, individual `<Tab />` elements will turn into `<Link />`. Keep in mind to add `href` attribute.

### HTML

In HTML world, after defininig appropriate `elements` prop in Controls panel (in Storybook) you can generate HTML snippet that will contain either `<button>` or `<a>` elements.

## Scrollable container

Tabs component is, by default, *scrollable* which means that user will be able to scroll through tabs once there's not enough horizontal space to fit them all.
Keep in mind that this should only be a case for mobile breakpoint. Because if you end up having scorllable tabs on desktop breakpoint then you should definitely talk to your designer.

## Configuration

### Stacked tabs

Tabs come in horizontal orientation by default. There's a setting for stacking them, one on top of another:
- For JSX it's just `stacked` prop.
- For HTML it's `c-tabs--stacked` class.

Keep in mind that stacked (vertical) tabs will turn into horizontal on the smallest breakpoint, automatically.

### Fitted layout

Tabs can take full available space (for 2 tabs: each will be 1/2 of the container, for 3 tabs: each will be 1/3 of the container, and so on). 
To turn this on:
- For JSX add `stacked` prop.
- For HTML add `c-tabs--fitted` class.

## Design
<iframe style="border: 1px solid rgba(0, 0, 0, 0.1);" width="100%" height="450" src="https://www.figma.com/embed?embed_host=share&url=https%3A%2F%2Fwww.figma.com%2Ffile%2FgHpwugeu9jFGW4d9u3ReuR%2FCrayons%3Fnode-id%3D1%253A141" allowfullscreen></iframe>
