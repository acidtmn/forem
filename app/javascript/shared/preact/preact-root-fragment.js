/**
 * A Preact 11+ implementation of the `replaceNode` parameter from Preact 10.
 *
 * This creates a "Persistent Fragment" (a fake DOM element) containing one or more
 * DOM nodes, which can then be passed as the `parent` argument to Preact's `render()` method.
 * Source: https://gist.github.com/developit/f4c67a2ede71dc2fab7f357f39cff28c
 */
export function createRootFragment(parent, replaceNode) {
  replaceNode = [].concat(replaceNode);
  var s = replaceNode[replaceNode.length-1].nextSibling;
  function insert(c, r) { parent.insertBefore(c, r || s); }
  return parent.__k = {
    nodeType: 1,
    parentNode: parent,
    firstChild: replaceNode[0],
    childNodes: replaceNode,
    insertBefore: insert,
    appendChild: insert,
    removeChild: function(c) { parent.removeChild(c); }
  };
}