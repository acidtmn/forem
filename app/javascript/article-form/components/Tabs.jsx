import { h } from 'preact';
import PropTypes from 'prop-types';
import { Button } from '@crayons';

export const Tabs = ({ onPreview, previewShowing }) => {
  return (
    <div className="crayons-article-form__tabs ml-auto">
      <Button
        variant="ghost"
        className={!previewShowing && 'current'}
        onClick={previewShowing && onPreview}
      >
        Edit
      </Button>
      <Button
        variant="ghost"
        className={previewShowing && 'current'}
        onClick={!previewShowing && onPreview}
      >
        Preview
      </Button>
    </div>
  );
};


Tabs.propTypes = {
  previewShowing: PropTypes.bool.isRequired,
  onPreview: PropTypes.func.isRequired,
};

Tabs.displayName = 'Tabs';
