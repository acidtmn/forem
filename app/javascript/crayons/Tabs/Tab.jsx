import { h } from 'preact';
import classNames from 'classnames/bind';
import { ButtonNew as Button, Link } from '@crayons';

export const Tab = (props) => {
  const { className, current, elements, fitted, ...otherProps } = props;

  const classes = classNames('c-tab', {
    'c-tab--fitted': fitted,
    [className]: className,
  });

  const sharedProps = {
    className: classes,
  };
  const buttonCurrentProps = current
    ? { 'aria-pressed': true }
    : { 'aria-pressed': false };
  const linkCurrentProps = current && { 'aria-current': 'page' };

  return elements === 'buttons' ? (
    <Button {...sharedProps} {...buttonCurrentProps} {...otherProps} />
  ) : (
    <Link block {...sharedProps} {...linkCurrentProps} {...otherProps} />
  );
};

Tab.displayName = 'Tab';
