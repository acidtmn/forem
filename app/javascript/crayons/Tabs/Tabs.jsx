import { h, cloneElement } from 'preact';
import classNames from 'classnames/bind';

export const Tabs = (props) => {
  const {
    children,
    stacked,
    fitted,
    className,
    elements = 'buttons',
    ...otherProps
  } = props;

  const Wrapper = elements === 'buttons' ? 'div' : 'nav';

  const classes = classNames('c-tabs', {
    'c-tabs--stacked': stacked,
    'c-tabs--fitted': fitted,
    [className]: className,
  });

  return (
    <Wrapper className={classes} {...otherProps}>
      {children.map((tab) => {
        return cloneElement(tab, { elements, fitted });
      })}
    </Wrapper>
  );
};

Tabs.displayName = 'Tabs';
