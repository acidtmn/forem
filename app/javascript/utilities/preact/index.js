import { render as preactRender } from 'preact';
import { unmountComponentAtNode } from 'preact/compat';

/**
 * Renders a Preact component into the given parent DOM element.
 *
 * Note: This function is calling Preact's render under the hood along with
 * a call to InstantClick's change event that explicitly unmounts the component rendered.
 *
 * @param {ComponentChild} vnode
 * @param {Element | Document | ShadowRoot | DocumentFragment} parent
 * @param {Element | Text} replaceNode
 */
export function render(vnode, parent, replaceNode) {
  InstantClick.on('change', () => {
    parent && unmountComponentAtNode(parent);
  });

  preactRender(vnode, parent, replaceNode);
}
