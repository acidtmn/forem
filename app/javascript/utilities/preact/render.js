import { render as preactRender } from 'preact';
import { unmountComponentAtNode } from 'preact/compat';

/**
 * Renders a Preact component into the given parent DOM element.
 *
 * Note: This function is calling Preact's render under the hood along with
 * a call to InstantClick's change event that explicitly unmounts the component rendered.
 *
 * @param {object} vnode The virtual DOM node representing a component.
 * @param {Element | Document | ShadowRoot | DocumentFragment} parent The DOM element where the component will render.
 * @param {Element | Text} replaceNode An optional DOM element that must of the given parent.
 * Instead of inferring where to start rendering, it will update or replace the passed element using Preact's diffing algorithm.
 */
export function render(vnode, parent, replaceNode) {
  InstantClick.on('change', () => {
    // We need to explicitly unmount a Preact component since we are using InstantClick. Failing to do so
    // would leave references to past components.
    parent && unmountComponentAtNode(parent);
  });

  preactRender(vnode, parent, replaceNode);
}
