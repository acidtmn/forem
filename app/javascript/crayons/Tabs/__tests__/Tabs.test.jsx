import { h } from 'preact';
import { render } from '@testing-library/preact';
import { axe } from 'jest-axe';
import { Tabs } from '../Tabs';
import { Tab } from '../Tab';

describe('<Tabs />', () => {
  it('has no accessibility errors when using buttons', async () => {
    const { container } = render(
      <Tabs elements="buttons">
        <Tab current>One</Tab>
        <Tab>two</Tab>
      </Tabs>,
    );

    const results = await axe(container);

    expect(results).toHaveNoViolations();
  });

  it('has no accessibility errors when using links', async () => {
    const { container } = render(
      <Tabs elements="links">
        <Tab current>One</Tab>
        <Tab>two</Tab>
      </Tabs>,
    );

    const results = await axe(container);

    expect(results).toHaveNoViolations();
  });

  it('renders default Tabs with buttons', () => {
    const { container } = render(
      <Tabs elements="buttons">
        <Tab current>One</Tab>
        <Tab>two</Tab>
      </Tabs>,
    );
    expect(container.innerHTML).toMatchSnapshot();
  });

  it('renders default Tabs with links', () => {
    const { container } = render(
      <Tabs elements="links">
        <Tab current>One</Tab>
        <Tab>two</Tab>
      </Tabs>,
    );
    expect(container.innerHTML).toMatchSnapshot();
  });

  it('renders stacked Tabs', () => {
    const { container } = render(
      <Tabs stacked>
        <Tab current>One</Tab>
        <Tab>two</Tab>
      </Tabs>,
    );

    expect(container.innerHTML).toMatchSnapshot();
  });

  it('renders fitted Tabs', () => {
    const { container } = render(
      <Tabs fitted>
        <Tab current>One</Tab>
        <Tab>two</Tab>
      </Tabs>,
    );

    expect(container.innerHTML).toMatchSnapshot();
  });
});
