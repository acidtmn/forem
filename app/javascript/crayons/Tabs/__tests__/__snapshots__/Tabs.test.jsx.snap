// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs /> renders default Tabs with buttons 1`] = `"<nav class=\\"c-tabs\\"><ul class=\\"c-tabs__list\\"><li class=\\"c-tabs__list__item\\"><button type=\\"button\\" class=\\"c-btn c-tab\\" aria-pressed=\\"true\\">One</button></li><li class=\\"c-tabs__list__item\\"><button type=\\"button\\" class=\\"c-btn c-tab\\" aria-pressed=\\"false\\">two</button></li></ul></nav>"`;

exports[`<Tabs /> renders default Tabs with links 1`] = `"<nav class=\\"c-tabs\\"><ul class=\\"c-tabs__list\\"><li class=\\"c-tabs__list__item\\"><a href=\\"#\\" class=\\"c-link c-link--block c-tab\\" aria-current=\\"page\\">One</a></li><li class=\\"c-tabs__list__item\\"><a href=\\"#\\" class=\\"c-link c-link--block c-tab\\">two</a></li></ul></nav>"`;

exports[`<Tabs /> renders fitted Tabs 1`] = `"<nav class=\\"c-tabs c-tabs--fitted\\"><ul class=\\"c-tabs__list\\"><li class=\\"c-tabs__list__item\\"><button type=\\"button\\" class=\\"c-btn c-tab c-tab--fitted\\" aria-pressed=\\"true\\">One</button></li><li class=\\"c-tabs__list__item\\"><button type=\\"button\\" class=\\"c-btn c-tab c-tab--fitted\\" aria-pressed=\\"false\\">two</button></li></ul></nav>"`;

exports[`<Tabs /> renders stacked Tabs 1`] = `"<nav class=\\"c-tabs c-tabs--stacked\\"><ul class=\\"c-tabs__list\\"><li class=\\"c-tabs__list__item\\"><button type=\\"button\\" class=\\"c-btn c-tab\\" aria-pressed=\\"true\\">One</button></li><li class=\\"c-tabs__list__item\\"><button type=\\"button\\" class=\\"c-btn c-tab\\" aria-pressed=\\"false\\">two</button></li></ul></nav>"`;
